okhttp/src/nonJvmMain/kotlin/okhttp3/internal/-HostnamesNonJvm.kt

```diff
@@ -16,8 +16,27 @@
 package okhttp3.internal
 
 import com.squareup.okhttpicu.SYSTEM_NORMALIZER
+import okhttp3.internal.idn.IDNA_MAPPING_TABLE
+import okio.Buffer
 
 internal actual fun idnToAscii(host: String): String {
-  // TODO implement properly
-  return SYSTEM_NORMALIZER.normalizeNfc(host).lowercase()
+  val bufferA = Buffer().writeUtf8(host)
+  val bufferB = Buffer()
+
+  // 1. Map, from bufferA to bufferB.
+  while (!bufferA.exhausted()) {
+    val codePoint = bufferA.readUtf8CodePoint()
```

Comment:
Can this fail by exhausting the string on bad input, is that caught in the common logic calling this?

Reply:
Good question!

This is all a hack to get around the fact that `kotlin.String` can’t iterate code points in multiplatform. (What a fail!)

This could throw an `EOFException` if the UTF-8 is malformed, but it won’t be 'cause we just encoded it above.

So really it’ll make forward progress thru the very last byte.